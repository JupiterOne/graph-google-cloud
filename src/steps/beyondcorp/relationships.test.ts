import {
  createMockStepExecutionContext,
  executeStepWithDependencies,
  filterGraphObjects,
  Recording,
  StepTestConfig,
} from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig, invocationConfig } from '../..';
import {
  BEYONDCORP_APP_CONNECTION_CLASS,
  BEYONDCORP_APP_CONNECTION_TYPE,
  BEYONDCORP_APP_CONNECTOR_CLASS,
  BEYONDCORP_APP_CONNECTOR_TYPE,
  BEYONDCORP_ENTERPRISE_CLASS,
  BEYONDCORP_ENTERPRISE_TYPE,
  RELATIONSHIP_TYPE_PROJECT_HAS_BEYONDCORP_ENTERPRISE,
  RELATIONSHIP_TYPE_PROJECT_USES_APP_CONNECTION,
  RELATIONSHIP_TYPE_PROJECT_USES_APP_CONNECTOR,
  STEP_APP_CONNECTION_HAS_APP_CONNECTOR_RELATIONSHIP,
  STEP_APP_CONNECTION_HAS_APPLICATION_ENDPOINT_RELATIONSHIP,
  STEP_APP_CONNECTION_HAS_GATEWAY_RELATIONSHIP,
  STEP_APPLICATION_ENDPOINT_USES_GATEWAY_RELATIONSHIP,
  STEP_PROJECT_HAS_BEYONDCORP_ENTERPRISE_RELATIONSHIP,
  STEP_PROJECT_USES_APP_CONNECTION_RELATIONSHIP,
  STEP_PROJECT_USES_APP_CONNECTOR_RELATIONSHIP,
} from './constant';
import {
  setupGoogleCloudRecording,
  getMatchRequestsBy,
} from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import {
  ExplicitRelationship,
  Relationship,
} from '@jupiterone/integration-sdk-core';
import {
  fetchResourceManagerProject,
  PROJECT_ENTITY_TYPE,
} from '../resource-manager';
import {
  buildProjectHasBeyondcorpEnterpriseRelationship,
  buildProjectUsesAppConnectorRelationship,
  fetchAppConnections,
  fetchAppConnectors,
  fetchBeyondcorpEnterprise,
} from '.';

describe(`beyondcorp#${STEP_APP_CONNECTION_HAS_APP_CONNECTOR_RELATIONSHIP}`, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });

  jest.setTimeout(45000);

  test(STEP_APP_CONNECTION_HAS_APP_CONNECTOR_RELATIONSHIP, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_APP_CONNECTION_HAS_APP_CONNECTOR_RELATIONSHIP,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfig),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_APP_CONNECTION_HAS_APP_CONNECTOR_RELATIONSHIP,
      instanceConfig: integrationConfig,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});

describe(`beyondcorp#${STEP_APP_CONNECTION_HAS_APPLICATION_ENDPOINT_RELATIONSHIP}`, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });

  jest.setTimeout(45000);

  test(STEP_APP_CONNECTION_HAS_APPLICATION_ENDPOINT_RELATIONSHIP, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_APP_CONNECTION_HAS_APPLICATION_ENDPOINT_RELATIONSHIP,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfig),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_APP_CONNECTION_HAS_APPLICATION_ENDPOINT_RELATIONSHIP,
      instanceConfig: integrationConfig,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});

describe(`beyondcorp#${STEP_APP_CONNECTION_HAS_GATEWAY_RELATIONSHIP}`, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });

  jest.setTimeout(45000);

  test(STEP_APP_CONNECTION_HAS_GATEWAY_RELATIONSHIP, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_APP_CONNECTION_HAS_GATEWAY_RELATIONSHIP,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfig),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_APP_CONNECTION_HAS_GATEWAY_RELATIONSHIP,
      instanceConfig: integrationConfig,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});

describe(`beyondcorp#${STEP_PROJECT_USES_APP_CONNECTOR_RELATIONSHIP}`, () => {
  let recording: Recording;
  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: STEP_PROJECT_USES_APP_CONNECTOR_RELATIONSHIP,
    });
  });
  afterEach(async () => {
    await recording.stop();
  });
  function separateRelationships(collectedRelationships: Relationship[]) {
    const { targets: directRelationships } = filterGraphObjects(
      collectedRelationships,
      (r) => !r._mapping,
    ) as {
      targets: ExplicitRelationship[];
    };
    return {
      directRelationships,
    };
  }
  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: {
        ...integrationConfig,
        serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
          'j1-gc-integration-dev-v2',
          'j1-gc-integration-dev-v3',
        ),
        serviceAccountKeyConfig: {
          ...integrationConfig.serviceAccountKeyConfig,
          project_id: 'j1-gc-integration-dev-v3',
        },
      },
    });
    await fetchAppConnectors(context);
    await fetchResourceManagerProject(context);
    await buildProjectUsesAppConnectorRelationship(context);
    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();
    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === PROJECT_ENTITY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_project' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          projectId: { type: 'string' },
          name: { type: 'string' },
          displayName: { type: 'string' },
          parent: { type: 'string' },
          lifecycleState: { type: 'string' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });
    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === BEYONDCORP_APP_CONNECTOR_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: BEYONDCORP_APP_CONNECTOR_CLASS,
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: BEYONDCORP_APP_CONNECTOR_TYPE },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          projectId: { type: 'string' },
          createdTime: { type: 'string' },
          updatedTime: { type: 'string' },
          UID: { type: 'string' },
          State: { type: 'string' },
          PrincipleInfo: { type: 'object' },
          CIDR: { type: 'string' },
          internal: { type: 'boolean' },
          public: { type: 'boolean' },
        },
      },
    });
    const { directRelationships } = separateRelationships(
      context.jobState.collectedRelationships,
    );
    expect(
      directRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_PROJECT_USES_APP_CONNECTOR,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'USES' },
          _type: {
            const: RELATIONSHIP_TYPE_PROJECT_USES_APP_CONNECTOR,
          },
        },
      },
    });
  });
});

describe(`beyondcorp#${STEP_PROJECT_USES_APP_CONNECTION_RELATIONSHIP}`, () => {
  let recording: Recording;
  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: STEP_PROJECT_USES_APP_CONNECTION_RELATIONSHIP,
    });
  });
  afterEach(async () => {
    await recording.stop();
  });
  function separateRelationships(collectedRelationships: Relationship[]) {
    const { targets: directRelationships } = filterGraphObjects(
      collectedRelationships,
      (r) => !r._mapping,
    ) as {
      targets: ExplicitRelationship[];
    };
    return {
      directRelationships,
    };
  }
  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: {
        ...integrationConfig,
        serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
          'j1-gc-integration-dev-v2',
          'j1-gc-integration-dev-v3',
        ),
        serviceAccountKeyConfig: {
          ...integrationConfig.serviceAccountKeyConfig,
          project_id: 'j1-gc-integration-dev-v3',
        },
      },
    });
    await fetchAppConnections(context);
    await fetchResourceManagerProject(context);
    await buildProjectUsesAppConnectorRelationship(context);
    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();
    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === PROJECT_ENTITY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_project' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          projectId: { type: 'string' },
          name: { type: 'string' },
          displayName: { type: 'string' },
          parent: { type: 'string' },
          lifecycleState: { type: 'string' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });
    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === BEYONDCORP_APP_CONNECTION_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: BEYONDCORP_APP_CONNECTION_CLASS,
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: BEYONDCORP_APP_CONNECTION_TYPE },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          createdTime: { type: 'string' },
          gateway: { type: 'string' },
          connector: { type: 'array' },
          uid: { type: 'string' },
          type: { type: 'string' },
          updatedTime: { type: 'string' },
          state: { type: 'string' },
          endpointHost: { type: 'string' },
          endpointPort: { type: 'number' },
          CIDR: { type: 'string' },
          internal: { type: 'boolean' },
          public: { type: 'boolean' },
        },
      },
    });
    const { directRelationships } = separateRelationships(
      context.jobState.collectedRelationships,
    );
    expect(
      directRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_PROJECT_USES_APP_CONNECTION,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'USES' },
          _type: {
            const: RELATIONSHIP_TYPE_PROJECT_USES_APP_CONNECTION,
          },
        },
      },
    });
  });
});

describe(`beyondcorp#${STEP_APPLICATION_ENDPOINT_USES_GATEWAY_RELATIONSHIP}`, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });

  jest.setTimeout(45000);

  test(STEP_APPLICATION_ENDPOINT_USES_GATEWAY_RELATIONSHIP, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_APPLICATION_ENDPOINT_USES_GATEWAY_RELATIONSHIP,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfig),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_APPLICATION_ENDPOINT_USES_GATEWAY_RELATIONSHIP,
      instanceConfig: integrationConfig,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});

describe(`beyondcorp#${STEP_PROJECT_HAS_BEYONDCORP_ENTERPRISE_RELATIONSHIP}`, () => {
  let recording: Recording;
  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: STEP_PROJECT_HAS_BEYONDCORP_ENTERPRISE_RELATIONSHIP,
    });
  });
  afterEach(async () => {
    await recording.stop();
  });
  function separateRelationships(collectedRelationships: Relationship[]) {
    const { targets: directRelationships } = filterGraphObjects(
      collectedRelationships,
      (r) => !r._mapping,
    ) as {
      targets: ExplicitRelationship[];
    };
    return {
      directRelationships,
    };
  }
  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: {
        ...integrationConfig,
        serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
          'j1-gc-integration-dev-v2',
          'j1-gc-integration-dev-v3',
        ),
        serviceAccountKeyConfig: {
          ...integrationConfig.serviceAccountKeyConfig,
          project_id: 'j1-gc-integration-dev-v3',
        },
      },
    });
    await fetchBeyondcorpEnterprise(context);
    await fetchResourceManagerProject(context);
    await buildProjectHasBeyondcorpEnterpriseRelationship(context);
    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();
    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === PROJECT_ENTITY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_project' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          projectId: { type: 'string' },
          name: { type: 'string' },
          displayName: { type: 'string' },
          parent: { type: 'string' },
          lifecycleState: { type: 'string' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });
    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === BEYONDCORP_ENTERPRISE_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: BEYONDCORP_ENTERPRISE_CLASS,
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: BEYONDCORP_ENTERPRISE_TYPE },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          category: { type: 'array' },
          function: { type: 'array' },
          endpoint: { type: 'string' },
        },
      },
    });
    const { directRelationships } = separateRelationships(
      context.jobState.collectedRelationships,
    );
    expect(
      directRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_PROJECT_HAS_BEYONDCORP_ENTERPRISE,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: RELATIONSHIP_TYPE_PROJECT_HAS_BEYONDCORP_ENTERPRISE,
          },
        },
      },
    });
  });
});
