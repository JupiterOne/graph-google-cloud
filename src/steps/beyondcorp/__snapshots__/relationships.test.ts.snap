// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`beyondcorp#fetch-project-has-beyondcorp-enterprise should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Service",
      ],
      "_key": "958457776463_google_cloud_beyondcorp_enterprise",
      "_rawData": [],
      "_type": "google_cloud_beyondcorp_enterprise",
      "category": [
        "Security",
        "Network",
      ],
      "createdOn": undefined,
      "displayName": "Beyondcorp Enterprise Service",
      "endpoint": "http://console.cloud.google.com/",
      "function": [
        "Networking",
      ],
      "name": "Beyondcorp Enterprise Service",
    },
    {
      "_class": [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "W/"8f56735f6f3d890d"",
            "labels": {
              "firebase": "enabled",
            },
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2024-05-15T11:48:40.174923Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "tag.firebase": "enabled",
      "updatedOn": 1715773720174,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "HAS",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|has|958457776463_google_cloud_beyondcorp_enterprise",
      "_toEntityKey": "958457776463_google_cloud_beyondcorp_enterprise",
      "_type": "google_cloud_project_has_beyondcorp_enterprise",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": [
    "google_cloud_beyondcorp_enterprise",
    "google_cloud_project",
    "google_cloud_project_has_beyondcorp_enterprise",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;

exports[`beyondcorp#fetch-project-uses-app-connection should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Network",
      ],
      "_key": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-connection-demo-3",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "applicationEndpoint": {
              "host": "localhost1",
              "port": 8000,
            },
            "connectors": [
              "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector",
            ],
            "createTime": "2024-04-08T12:30:06.446139022Z",
            "displayName": "connection-display-name-2",
            "gateway": {
              "appGateway": "projects/j1-gc-integration-dev-v3/locations/us-central1/appGateways/autogenerated-gateway-1",
              "ingressPort": 19445,
              "uri": "https://www.googleapis.com/compute/v1/projects/ff55b02a070816266p-tp/regions/us-central1/serviceAttachments/bce-ps7-sj2irxmhno1tdihn-1-slm0-2",
            },
            "labels": {
              "foo": "bar",
            },
            "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-connection-demo-3",
            "state": "CREATED",
            "type": "TCP_PROXY",
            "uid": "9ec0aaa4-8376-4101-92d3-47035b940b99",
            "updateTime": "2024-04-08T12:30:31.251527880Z",
          },
        },
      ],
      "_type": "google_cloud_beyondcorp_connection",
      "connector": [
        "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector",
      ],
      "createdOn": undefined,
      "createdTime": "2024-04-08T12:30:06.446139022Z",
      "displayName": "connection-display-name-2",
      "endpointHost": "localhost1",
      "endpointPort": 8000,
      "gateway": "projects/j1-gc-integration-dev-v3/locations/us-central1/appGateways/autogenerated-gateway-1",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-connection-demo-3",
      "state": "CREATED",
      "tag.foo": "bar",
      "type": "TCP_PROXY",
      "uid": "9ec0aaa4-8376-4101-92d3-47035b940b99",
      "updatedTime": "2024-04-08T12:30:31.251527880Z",
    },
    {
      "_class": [
        "Network",
      ],
      "_key": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-demo-connection",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "applicationEndpoint": {
              "host": "localhost",
              "port": 8080,
            },
            "createTime": "2024-04-08T07:20:40.432467811Z",
            "gateway": {
              "appGateway": "projects/j1-gc-integration-dev-v3/locations/us-central1/appGateways/autogenerated-gateway-1",
              "ingressPort": 19443,
              "uri": "https://www.googleapis.com/compute/v1/projects/ff55b02a070816266p-tp/regions/us-central1/serviceAttachments/bce-ps7-sj2irxmhno1tdihn-1-slm0-0",
            },
            "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-demo-connection",
            "state": "CREATED",
            "type": "TCP_PROXY",
            "uid": "4775e0ff-53d7-418b-afa9-e7c618585350",
            "updateTime": "2024-04-08T07:26:42.546018228Z",
          },
        },
      ],
      "_type": "google_cloud_beyondcorp_connection",
      "connector": undefined,
      "createdOn": undefined,
      "createdTime": "2024-04-08T07:20:40.432467811Z",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-demo-connection",
      "endpointHost": "localhost",
      "endpointPort": 8080,
      "gateway": "projects/j1-gc-integration-dev-v3/locations/us-central1/appGateways/autogenerated-gateway-1",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-demo-connection",
      "state": "CREATED",
      "type": "TCP_PROXY",
      "uid": "4775e0ff-53d7-418b-afa9-e7c618585350",
      "updatedTime": "2024-04-08T07:26:42.546018228Z",
    },
    {
      "_class": [
        "Network",
      ],
      "_key": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-connection-demo2",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "applicationEndpoint": {
              "host": "localhost",
              "port": 8080,
            },
            "connectors": [
              "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector-2demo",
            ],
            "createTime": "2024-04-08T09:37:19.488886543Z",
            "displayName": "connection-display-name",
            "gateway": {
              "appGateway": "projects/j1-gc-integration-dev-v3/locations/us-central1/appGateways/autogenerated-gateway-1",
              "ingressPort": 19444,
              "uri": "https://www.googleapis.com/compute/v1/projects/ff55b02a070816266p-tp/regions/us-central1/serviceAttachments/bce-ps7-sj2irxmhno1tdihn-1-slm0-1",
            },
            "labels": {
              "foo": "bar",
            },
            "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-connection-demo2",
            "state": "CREATED",
            "type": "TCP_PROXY",
            "uid": "85dab80a-57d5-4768-8591-99aa23b89b8f",
            "updateTime": "2024-04-08T09:38:56.612964882Z",
          },
        },
      ],
      "_type": "google_cloud_beyondcorp_connection",
      "connector": [
        "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector-2demo",
      ],
      "createdOn": undefined,
      "createdTime": "2024-04-08T09:37:19.488886543Z",
      "displayName": "connection-display-name",
      "endpointHost": "localhost",
      "endpointPort": 8080,
      "gateway": "projects/j1-gc-integration-dev-v3/locations/us-central1/appGateways/autogenerated-gateway-1",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnections/my-connection-demo2",
      "state": "CREATED",
      "tag.foo": "bar",
      "type": "TCP_PROXY",
      "uid": "85dab80a-57d5-4768-8591-99aa23b89b8f",
      "updatedTime": "2024-04-08T09:38:56.612964882Z",
    },
    {
      "_class": [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "W/"8f56735f6f3d890d"",
            "labels": {
              "firebase": "enabled",
            },
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2024-05-15T11:48:40.174923Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "tag.firebase": "enabled",
      "updatedOn": 1715773720174,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "google_cloud_beyondcorp_connection",
    "google_cloud_project",
  ],
  "numCollectedEntities": 4,
  "numCollectedRelationships": 0,
}
`;

exports[`beyondcorp#fetch-project-uses-app-connector should collect data 1`] = `
{
  "collectedEntities": [
    {
      "State": "CREATED",
      "UID": "9ef8f600-6666-4afa-b2da-4b246e788309",
      "_class": [
        "Network",
      ],
      "_key": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "createTime": "2024-04-05T07:28:27.484117436Z",
            "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector",
            "principalInfo": {
              "serviceAccount": {
                "email": "j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com",
              },
            },
            "state": "CREATED",
            "uid": "9ef8f600-6666-4afa-b2da-4b246e788309",
            "updateTime": "2024-04-05T07:29:12.033754074Z",
          },
        },
      ],
      "_type": "google_cloud_beyondcorp_connector",
      "createdOn": undefined,
      "createdTime": "2024-04-05T07:28:27.484117436Z",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector",
      "updatedTime": "2024-04-05T07:29:12.033754074Z",
    },
    {
      "State": "CREATED",
      "UID": "c7126d9e-3fad-4ffb-a024-df44918ca97d",
      "_class": [
        "Network",
      ],
      "_key": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector-2demo",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "createTime": "2024-04-05T10:30:24.351163994Z",
            "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector-2demo",
            "principalInfo": {
              "serviceAccount": {
                "email": "j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com",
              },
            },
            "state": "CREATED",
            "uid": "c7126d9e-3fad-4ffb-a024-df44918ca97d",
            "updateTime": "2024-04-05T10:31:00.015811832Z",
          },
        },
      ],
      "_type": "google_cloud_beyondcorp_connector",
      "createdOn": undefined,
      "createdTime": "2024-04-05T10:30:24.351163994Z",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector-2demo",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector-2demo",
      "updatedTime": "2024-04-05T10:31:00.015811832Z",
    },
    {
      "State": "CREATED",
      "UID": "7429b1ce-08ed-490f-888f-d8ddd1bcf849",
      "_class": [
        "Network",
      ],
      "_key": "projects/j1-gc-integration-dev-v3/locations/us-east1/appConnectors/my-connector-demo3",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "createTime": "2024-04-08T09:07:18.868443636Z",
            "name": "projects/j1-gc-integration-dev-v3/locations/us-east1/appConnectors/my-connector-demo3",
            "principalInfo": {
              "serviceAccount": {
                "email": "j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com",
              },
            },
            "state": "CREATED",
            "uid": "7429b1ce-08ed-490f-888f-d8ddd1bcf849",
            "updateTime": "2024-04-08T09:07:52.614057708Z",
          },
        },
      ],
      "_type": "google_cloud_beyondcorp_connector",
      "createdOn": undefined,
      "createdTime": "2024-04-08T09:07:18.868443636Z",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-east1/appConnectors/my-connector-demo3",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-east1/appConnectors/my-connector-demo3",
      "updatedTime": "2024-04-08T09:07:52.614057708Z",
    },
    {
      "_class": [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "W/"8f56735f6f3d890d"",
            "labels": {
              "firebase": "enabled",
            },
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2024-05-15T11:48:40.174923Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "tag.firebase": "enabled",
      "updatedOn": 1715773720174,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "USES",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|uses|projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector",
      "_toEntityKey": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector",
      "_type": "google_cloud_project_uses_beyondcorp_connector",
      "displayName": "USES",
    },
    {
      "_class": "USES",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|uses|projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector-2demo",
      "_toEntityKey": "projects/j1-gc-integration-dev-v3/locations/us-central1/appConnectors/my-connector-2demo",
      "_type": "google_cloud_project_uses_beyondcorp_connector",
      "displayName": "USES",
    },
    {
      "_class": "USES",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|uses|projects/j1-gc-integration-dev-v3/locations/us-east1/appConnectors/my-connector-demo3",
      "_toEntityKey": "projects/j1-gc-integration-dev-v3/locations/us-east1/appConnectors/my-connector-demo3",
      "_type": "google_cloud_project_uses_beyondcorp_connector",
      "displayName": "USES",
    },
  ],
  "encounteredTypes": [
    "google_cloud_beyondcorp_connector",
    "google_cloud_project",
    "google_cloud_project_uses_beyondcorp_connector",
  ],
  "numCollectedEntities": 4,
  "numCollectedRelationships": 3,
}
`;
